test/opt/compact_ids_test.cpp

```diff
@@ -87,4 +90,57 @@ OpFunctionEnd
   SinglePassRunAndCheck<opt::CompactIdsPass>(before, after, false, false);
 }
 
+TEST(CompactIds, InstructionResultIsUpdated) {
+  // For https://github.com/KhronosGroup/SPIRV-Tools/issues/827
+  // In that bug, the compact Ids pass was directly updating the result Id
+  // word for an OpFunction instruction, but not updating the cached
+  // result_id_ in that Instruction object.
+  //
+  // This test is a bit cheesy.  We don't expose internal interfaces enough
+  // to see the inconsistency.  So reproduce the original scenario, with
+  // compact ids followed by a pass that trips up on the inconsistency.
+
+  const std::string input(R"(OpCapability Shader
+OpMemoryModel Logical Simple
+OpEntryPoint GLCompute %100 "main"
+%200 = OpTypeVoid
+%300 = OpTypeFunction %200
+%100 = OpFunction %300 None %200
+%400 = OpLabel
+OpReturn
+OpFunctionEnd
+)");
+
+  std::vector<uint32_t> binary;
+  const spv_target_env env = SPV_ENV_UNIVERSAL_1_0;
+  spvtools::SpirvTools tools(env);
+  auto assembled = tools.Assemble(
+      input, &binary, SPV_TEXT_TO_BINARY_OPTION_PRESERVE_NUMERIC_IDS);
+  EXPECT_TRUE(assembled);
+
+  spvtools::Optimizer optimizer(env);
+  optimizer.RegisterPass(CreateCompactIdsPass());
+  // The exhaustive inliner will use the result_id
+  optimizer.RegisterPass(CreateInlineExhaustivePass());
+
+  // This should not crash!
```

Comment:
A badge of code quality...

Reply:
Yeah.  I debated on the "proper" way to test this.  First thought was to expose the Instruction objects directly, but that would break the current information-hiding.  That's generally ok, but we've got plans to overhaul the instruction classes and didn't want to cause churn at this moment.